### DISCUSSION TITLE
Could charm.sh enable the dagger CLI to generate an ASCII diagram for the DAG?

### DISCUSSION THREAD

**Discussion by gerhard** [dagger/dagger#1587] [Ideas]
https://charm.sh/ looks interesting, no idea if it will work. Capturing it as a thought.

The idea is for the `dagger` CLI to be able to generate something along the lines of:

```
                 app
                 |
                 v
test_db_start    deps ----------------------------------------\
|                |                      |                     |
|                v                      v                     v
|                deps_compile_test      deps_compile_prod     assets_compile
|                |                      |               |     |
|                |                      |               \---->|
|                v                      v                     |
\--------------> test                   image_prod_cache      assets_digest
                 |                      |                     |
test_db_stop <---|                      v                     |
                 |                      image_prod <----------/
                 |                      |
                 |                      v
                 \--------------------> image_prod_tag
```

There is more context from Toby Padilla, one of the Charm co-founders, in this [Changelog Slack thread](https://changelog.slack.com/archives/C024Q4CER/p1644423452811479) (see the comments below for more details.

Initial idea & context: https://github.com/thechangelog/changelog.com/pull/395

**Comment by gurjeet** [comment]
I don't see anything related to ASCII diagrams on charm.sh website; maybe I'm missing some context.

Maybe this would help:

ascii-graphs
Parse ASCII art diagrams of graphs (nodes and edges) 
https://github.com/mdr/ascii-graphs

Or just plain old Grpahviz, compiled to JS/WASM: http://viz-js.com/ (not maintained).

Or the more modern, and actively developed MermaidJS: https://mermaid-js.github.io/

**Reply by gerhard** [reply]
You are right, this does not exist today. Context & [link to original Changelog Slack thread](https://changelog.slack.com/archives/C024Q4CER/p1644423452811479):

<img width="398" alt="image" src="https://user-images.githubusercontent.com/3342/153389598-9d96de3f-3073-49ae-ae07-cf8ba43b6bc8.png">

**Reply by gerhard** [reply]
https://github.com/mdr/ascii-graphs comes close, but it's not Go.

We would need something in Go so that dagger CLI could generate something along the lines of https://github.com/thechangelog/changelog.com/blob/0b5e1a7999a8179a5d892a1974e2175a9acf37fa/2021/dagger/prod_image/main.cue#L45-L63

More context: https://github.com/thechangelog/changelog.com/pull/395

**Comment by grouville** [comment]
As it doesn't exist atm [Gographviz creates an image from it, it's not ascii (https://github.com/goccy/go-graphviz)], one solution would be to rely on that: https://github.com/ggerganov/dot-to-ascii

**Reply by gerhard** [reply]
Nice, I'll mention this in the original Slack thread.

**Comment by gurjeet** [comment]
The [dot-to-ascii][] is an online webapp.  It's a PHP script that receives input from client, invoke a Perl script, and returns the result to the client. I think this is too heavy, fragile, and insecure a tool to be used and relied upon by dagger cli.

I feel [go-graphviz][] has a good chance of implementing ascii output, in addition to the many formats it already does. And it would be the appropriate project for a CLI to rely upon; local execution removes security, fragility, and load concerns.

[dot-to-ascii]: https://github.com/ggerganov/dot-to-ascii
[go-graphviz]: https://github.com/goccy/go-graphviz

**Reply by grouville** [reply]
Done: https://github.com/goccy/go-graphviz/issues/53 🤞

**Comment by gurjeet** [comment]
The Perl script that dot-to-ascii relies on, seems to be a long-standing package Graph::Easy, widely used, widely recommended, and well-maintained.

Since the daggers CLI expects buildkit to be present, would it be an option to run that in a container and grab the output. It'd still be local execution, with the other attributes of stability and security, but it'd be heavier compared using a Golang package.

**Reply by gurjeet** [reply]
s/daggers/dagger/

**Comment by jlongtine** [comment]
@gurjeet @grouville I looked at `Graph::Easy` a bit last night. I'm super impressed with how well it does. The unicode output in particular looks very good. Generating the DAG at each step of evaluation should be relatively straightforward. And we _might_ be able to utilize buildkit to run `Graph::Easy` in a container. Although part of me just wants to figure out how to extract the `dot` -> `ascii/unicode` bits of `Graph::Easy` and rewrite them in golang. That'd be awesome. :)

**Reply by grouville** [reply]
If at some point you need some hands, I'll be glad to help 😇

**Reply by jlongtine** [reply]
This is how `Graph::Easy` renders [Go Package Imports | Graphviz](https://graphviz.org/Gallery/directed/go-package.html) 

![Go Package Imports](https://user-images.githubusercontent.com/22073/153929023-c8fbbb0c-d56e-4175-8fcd-82b3af85b9c4.png)

Rather impressive. Especially given the medium.

**Reply by jlongtine** [reply]
Here's a URL dump of interesting/potentially useful things: 

- [Graph::Easy - Brave Search](https://search.brave.com/search?q=Graph%3A%3AEasy&source=desktop)
- [Graph::Easy - Manual - Overview](http://bloodgate.com/perl/graph/manual/overview.html)
- [Graph::Easy - Convert or render graphs (as ASCII, HTML, SVG or via Graphviz) - metacpan.org](https://metacpan.org/pod/Graph::Easy)
- [graph-easy(1) - Linux man page](https://linux.die.net/man/1/graph-easy)
- [gitpan/Graph-Easy: Read-only release history for Graph-Easy](https://github.com/gitpan/Graph-Easy)
- [Browse SHLOMIF/Graph-Easy-0.76 - metacpan.org](https://metacpan.org/release/SHLOMIF/Graph-Easy-0.76/source)
- [Issues · goccy/go-graphviz](https://github.com/goccy/go-graphviz/issues?q=is%3Aissue+is%3Aclosed)
- [graphviz golang - Brave Search](https://search.brave.com/search?q=graphviz+golang&source=desktop)
- [gographviz package - github.com/awalterschulze/gographviz - pkg.go.dev](https://pkg.go.dev/github.com/awalterschulze/gographviz#ParseString)
- [concentrate | Graphviz](https://graphviz.gitlab.io/docs/attrs/concentrate/)
- [Graphviz Visual Editor](http://magjac.com/graphviz-visual-editor/)
- [Go, Graphs and Graphviz – tpaschalis – software, systems](https://tpaschalis.github.io/golang-graphviz/)
- [Converts 'go mod graph' output into Graphviz's DOT language | GolangRepo](https://golangrepo.com/repo/lucasepe-modgv-go-package-management)
- [gographviz package - github.com/awalterschulze/gographviz - pkg.go.dev](https://pkg.go.dev/github.com/awalterschulze/gographviz#NewGraph)
- [emicklei/dot: Go package for writing descriptions using the Graphviz DOT language](https://github.com/emicklei/dot/)
- [lucasepe/dot: Go package for writing Graph descriptions using the Graphviz DOT language](https://github.com/lucasepe/dot)
- [graph layout golang - Brave Search](https://search.brave.com/search?q=graph+layout+golang&source=desktop)
- [JohnCoene/go-nettle: Basic Graph Layout](https://github.com/JohnCoene/go-nettle)
- [layout package - gonum.org/v1/gonum/graph/layout - pkg.go.dev](https://pkg.go.dev/gonum.org/v1/gonum@v0.9.3/graph/layout#NewOptimizerR2)
- [layout package - gonum.org/v1/gonum/graph/layout - pkg.go.dev](https://pkg.go.dev/gonum.org/v1/gonum@v0.9.3/graph/layout#OptimizerR2)
- [graph package - gonum.org/v1/gonum/graph - pkg.go.dev](https://pkg.go.dev/gonum.org/v1/gonum@v0.9.3/graph#Graph)
- [graph package - gonum.org/v1/gonum/graph - pkg.go.dev](https://pkg.go.dev/gonum.org/v1/gonum@v0.9.3/graph#Node)
- [Box Drawing — Unicode Character Table](https://jrgraphix.net/r/Unicode/2500-257F)
- [Go Package Imports | Graphviz](https://graphviz.org/Gallery/directed/go-package.html)

**Reply by StevenACoffman** [reply]
Hi @jlongtine If you are looking for directed acyclic graph layout algorithms in pure Go, https://github.com/nikolaydubina/go-graph-layout is currently the best, as it implements the Kozo Sugiyama layer algorithm, which does edge crossing minimization fairly efficiently. GraphViz is the gold standard, but requires cgo, as does the rearchitected [dynadag](https://www.dynagraph.org/) in C++. The closest port of GraphViz to another language besides Go is currently https://github.com/erikbrinkman/d3-dag/

**Reply by jlongtine** [reply]
@StevenACoffman Thanks for the links! [go-graph-layout](https://github.com/nikolaydubina/go-graph-layout) looks nice! The other piece that I'm still unsure how to do best is the actual rendering of ASCII/Unicode in the terminal given the node locations, etc. I haven't had the time to spend to dig more deeply into that.

**Comment by verdverm** [comment]
The harder part will be dynamic tasks, created by comprehensions, possibly based on the output of previous tasks. These cannot be known by simply reading the input CUE.

**Reply by jlongtine** [reply]
@verdverm Totally agree here. I don't know that there is any straightforward way to get dynamic tasks. It _might_ possible with a walking of the cue values, but I don't think the reference (and thus dependency) information would be easy to get at.
